beartype/typing.py

```diff
@@ -288,3 +288,211 @@
+            # it from being considered part of the Protocol. (See
+            # <https://github.com/python/cpython/blob/main/Lib/typing.py>.)
+            cls._abc_inst_check_cache = {}
+
+            return cls
+
+        def __instancecheck__(cls, inst: Any) -> bool:
+            try:
+                # This has to stay *super* tight! Even adding a mere assertion
+                # can add ~50% to the best case runtime!
+                return cls._abc_inst_check_cache[type(inst)]
+            except KeyError:
+                # If you're going to do *anything*, do it here. Try not to
+                # expand the rest of this method if you can avoid it.
+                inst_t = type(inst)
+                bases_pass_muster = True
+
+                for base in cls.__bases__:
+                    if base is cls or base.__name__ in (
+                        "Protocol",
+                        "Generic",
+                        "object",
+                    ):
+                        continue
+
+                    if not isinstance(inst, base):
+                        bases_pass_muster = False
+                        break
+
+                cls._abc_inst_check_cache[
+                    inst_t
+                ] = bases_pass_muster and cls._check_only_my_attrs(inst)
+
+                return cls._abc_inst_check_cache[inst_t]
+
+        def _check_only_my_attrs(cls, inst: Any) -> bool:
+            attrs = set(cls.__dict__)
+            attrs.update(cls.__dict__.get("__annotations__", {}))
+            attrs.intersection_update(_get_protocol_attrs(cls))
+
+            for attr in attrs:
+                if not hasattr(inst, attr):
+                    return False
+                elif callable(getattr(cls, attr, None)) and getattr(inst, attr) is None:
+                    return False
+
+            return True
+
+    @runtime_checkable
+    class Protocol(_Protocol, Generic, metaclass=_CachingProtocolMeta):  # type: ignore [no-redef]
+        """
+        ``@beartype``-compatible (i.e., ``@runtime_checkable``) drop-in
+        replacement for :class:`typing.Protocol` that can lead to significant
+        performance improvements. Uses :class:`_CachingProtocolMeta` to cache
+        :func:`isinstance` check results.
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from abc import abstractmethod
+          >>> from beartype import beartype
+          >>> from beartype.typing import Protocol
+
+          >>> class MyBearProtocol(Protocol):  # <-- runtime-checkable through inheritance
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+
+          >>> @beartype
+          ... def do_somthing(thing: MyBearProtocol) -> None:
+          ...   thing.myfunc(0)
+        """
+        __slots__: Union[str, Iterable[str]] = ()
+
```

Comment:
These live in `_protocol.py` in the latest commit, but the issue of whether to include them at all remains.

Reply:
If we're going through the trouble of creating a drop-in caching replacement for `Protocol`, I'm going to assume we'll be successful and also assume that we would want caching versions of the standard library `Supports` protocols to go with it (and maybe even others).

It would seem silly to me to require a client who went through the trouble of typing `from beartype.typing import Protocol, SupportsIndex` to do _extra_ work to ensure the `SupportsIndex` they just imported got the benefits of the `Protocol` they just imported.

Resolving for now, but we can revisit if needed.